Cycle1 experiment refactor

mhoover opened the pull request:
This PR refactors the `NGS2_WITNESS_Cycle1_experiment_prep.R` script in a number of ways:
1. The script was refactored top-to-bottom to be able to process multiple experiment sessions. This involved refactoring code to automatically identify any raw input files in directories and process them (meaning people need to be careful what files are going into the input directories). 
2. It adds code and error-handling for experiment 1 when the network is static. Because ties don't change under a static configuration, there are no `rewiringEvents` which means part of the processing otherwise would break without some programming around it.
3. The script became too big and was serving too many purposes, so it was broken out to two scripts, one for experiment 1 (`NGS2_WITNESS_Cycle1_experiment1_prep.R`) and another for experiment 2 (`NGS2_WITNESS_Cycle1_experiment2_prep.R`). 

**This PR will have merge conflicts with #9 and so we should coordinate on merging before moving ahead.**

mhoover committed:
Start prep work for the first cycle of experiments.

mhoover committed:
Refactor for multiple experiments.

mhoover committed:
Finish refactor for experiment 1, started experiment 2.

mhoover committed:
Completed initial refactoring.

mhoover committed:
Refactor experimental processing scripts.

This commit breaks the `NGS2_WITNESS_Cycle1_experiment_prep.R`
script into two files, one for experiment 1 and another for
experiment 2 as together it was too unwieldy.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment1_prep.R:
```diff
@@ -0,0 +1,422 @@
+    lapply(list.of.packages, require, character.only = TRUE)
+
+    pacman::p_load(multiwayvcov, lmtest)
+}
```
Can we remove lines 3 to 14, as packages are managed from the "Master" script?

mhoover replied:
I'd rather not. I don't `source()` the master script, so pulling in libraries doesn't happen for me there. Given the `if(Sys.info()['sysname'] == 'Darwin') {` we should be fine -- that'll only execute this block if someone is using a Mac, so anyone running from a Gallup server will skip over this entirely.

C-Pablo-Diego-Rosell replied:
Hmmm, I'm betraying my PC-centrism here, but... is there a way to pull all the libraries from a central script, rather than having library calls on each separate script?

mhoover replied:
Yeah, the way you have it in `NGS2_WITNESS_Cycle1_MASTER.R` is a great way to do it... but I don't run the entire pipeline each time I process the Breadboard data.

C-Pablo-Diego-Rosell replied:
How about I pull the library calls into a separate script that does only that?

mhoover replied:
We're bikeshedding at this point. However you want it handled will be fine; I'll adjust accordingly. Lay out your plan here and I'll adjust the PR.

C-Pablo-Diego-Rosell replied:
OK, let us ignore this for now.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment1_prep.R:
```diff
@@ -0,0 +1,422 @@
+    x$condition <- ifelse(exp_cond == 0, 'Static',
+                          ifelse(exp_cond == .1, 'Viscous',
+                          ifelse(exp_cond == .3, 'Fluid',
+                          ifelse(exp_cond == 1, 'Random', 'Other'))))
```
Do we know for sure that the "Random" condition is given by "exp_cond == 1", or do we have to wait until seeing an actual "random" session?

mhoover replied:
Yeah, Mark confirmed this. `1` will indicate random... it just hasn't come up yet.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment1_prep.R:
```diff
@@ -0,0 +1,422 @@
+    }))
+}
+
+get_neighbor_number <- function(d, cnx, rd) {
```
Some of these functions are hard for me to interpret, but I am not sure that we are counting the number of neighbors right. For example, for in experiment "e1.1_2017-08-21_9250.csv", for player pid "3j95iGEm" I only see one "Connected" event at the beginning of round 1 (event id 1831602), and yet the processed file shows 6 neighbors in round 1. Also, there are some examples where "num_neighbors" is greater than "group_size". See for example session "e1.2_2017-08-24", pid "8xSP33VT", round 8, showing 15 neighbors, and a group size of 12.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment2_prep.R:
```diff
@@ -0,0 +1,257 @@
+    lapply(list.of.packages, require, character.only = TRUE)
+
+    pacman::p_load(multiwayvcov, lmtest)
+}
```
As above, can we delete lines 3-14?

mhoover replied:
So, I'd like to keep this as above, however, I don't actually need all these packages, so that can tidy that up.

C-Pablo-Diego-Rosell replied:
As above

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment2_prep.R:
```diff
@@ -0,0 +1,257 @@
+    return(tmp)
+})
+
+# Since group changes only happen at rounds 1, 4, 8, 12 and so on, need to
```
This is in fact not the case in the data. I see group change rounds at rounds 0, 1, 2, 3, 6, etc. all over the place, really. Also within the same session I see group change rounds varying across players.  We need to clarify with Mark/Feng.

mhoover replied:
Alright; this is definitely something we'll need to get clarified. This is code I left basically untouched... seems like changes can happen at random during rounds? If so, this block of code will need a definite change, since these breaks are hard-coded.

C-Pablo-Diego-Rosell replied:
Mark confirmed that individuals are offered a group change round at random, with a 25% chance every round, independently for every individual.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment2_prep.R:
```diff
@@ -0,0 +1,257 @@
+    return(x)
+})
+
+cooperation <- mapply(function(x, y) {
```
I believe this mapply is creating duplicated cooperation events. See for example in the resulting "cooperation_exp2.csv" file, playerid "8IV3mrD4" shows the exact same data twice for round_num ==1.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment2_prep.R:
```diff
@@ -0,0 +1,257 @@
+coop.events <- lapply(coop.events, function(x) {
+    x$recround <- cut(
+        x$data.value.curRound,
+        breaks = c(seq(1,24, 4), 24),
```
Once we clarify with Feng/Mark, we need to fix these breaks, the declared sequence "c(seq(1,24, 4), 24)" actually produces the following group change rounds: 1, 5,  9, 13, 17, 21, 24, rather than 1, 4, 8, 12, 16, 20, 24 as expected. This is something you inherited from earlier messy code.

mhoover replied:
Agree on the fixup.

C-Pablo-Diego-Rosell reviewed (changes requested):
Hi Matt, good job refactoring the data to take in multiple experiment instances. There are a few things to fix, most of them are probably issues you inherited from earlier messy code. For exp 1, I think the number of neighbors is not being calculated correctly. For experiment 2, the main problem is with the design of the game, with group change rounds all over the place. We need to clarify with Feng and Mark before fixing.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment1_prep.R:
```diff
@@ -0,0 +1,422 @@
+
+# add in empanelment id for non-bots
+coop <- merge_with_empanelment(coop, empanel_bb_xwalk, 'pid')
+
```
Can we drop useless sessions here? I suggest the following code. 

- Drop failed "e1.1_2017-08-21" session, including "other" conditions

coop<- subset (coop, condition!="Other")

- Drop tiny "e1.4_2017-08-31" session, with only three players

coop <- subset (coop, session!="e1.4_2017-08-31")

- Drop short "e1.3_2017-08-23" session, with only one round

coop <- subset (coop, session!="e1.3_2017-08-23")

mhoover replied:
Happy to drop bad sessions, but would like to make it condition-based. The `condition == 'Other'` is fine, but for the other, it would be good to have some cutoffs... otherwise, we'll have to go through each experiment to determine if it should be dropped. That is, what's the minimum number of rounds and the minimum number of players you'll take as floors? From there, I'll work in the code to adjust based on those values.

C-Pablo-Diego-Rosell replied:
Minimum number of rounds = 2
Minimum number of players = 10

mhoover replied:
This has been addressed with commit 8e90ef.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment1_prep.R:
```diff
@@ -0,0 +1,422 @@
+                          ifelse(exp_cond == 1, 'Random', 'Other'))))
+    return(x)
+}, coop, as.list(names(coop)), exp1, SIMPLIFY = FALSE))
+
```
- Please add "fluid_dummy" variable: 

exp1_cooperation$fluid_dummy[exp1_cooperation$condition=="Fluid"] <- 1
exp1_cooperation$fluid_dummy[exp1_cooperation$condition=="Random" |
                             exp1_cooperation$condition=="Viscous" |
                               exp1_cooperation$condition=="Static"] <- 0

mhoover replied:
You got it.

C-Pablo-Diego-Rosell commented on NGS2_WITNESS_Cycle1_experiment1_prep.R:
```diff
@@ -0,0 +1,422 @@
+                                 rename = TRUE, rename_var = 'player_empanel_id')
+rewire <- merge_with_empanelment(rewire, empanel_bb_xwalk, 'alterid',
+                                 rename = TRUE, rename_var = 'alter_empanel_id')
+
```
- Drop failed "e1.1_2017-08-21" session, including "other" conditions

rewire <- subset (rewire, session!="e1.1_2017-08-21")

- Drop tiny "e1.4_2017-08-31" session, with only three players

rewire <- subset (rewire, session!="e1.4_2017-08-31")

- Drop short "e1.3_2017-08-23" session, with only one round

rewire <- subset (rewire, session!="e1.3_2017-08-23")

mhoover replied:
See comment above -- I'm okay with this, as long as we make it condition-based.

C-Pablo-Diego-Rosell replied:
Minimum number of rounds = 2
Minimum number of players = 10

mhoover replied:
This has been addressed with commit 8e90ef.

mhoover committed:
Refactor experiment 2.

This commit handles the random probability chance for group
change that can happen after rounds.

mhoover committed:
Finish refactor and changes for experiment 1.

This commit addresses the concerns around how participants are
tied to one another. Before, ties were getting double-counted
in many cases, due to how data was represented in Breadboard. The
commit deals with this and also addresses destroying ties when
someone leaves an experiment. Additionally, it drops experiments
that don't meet certain criteria in terms of number of players or
rounds.
